### ISSUE TITLE
Add overloads to enumeration APIs that take FindOptions flags

### ISSUE THREAD

**Issue by JeremyKuhne** [dotnet/corefx#25875]
This is contingent on https://github.com/dotnet/corefx/issues/25873. If that API proposal is done, we should also expose overloads on the existing APIs that allow you to get at the new options.

## Rationale and Usage

Customers have been asking for the ability to tweak enumeration for common scenarios. Our new extensibility proposal (https://github.com/dotnet/corefx/issues/25873) has a new set of `FindOptions` flags that we should support to get:

- Recursion
- Ability to skip inaccessible files
- Ability to specify larger buffers for remote access
- No locking optimization for thread-bound enumeration

## Proposed API

As proposed in https://github.com/dotnet/corefx/issues/25873, the Flags would be:

``` C#
    [Flags]
    public enum FindOptions
    {
        None = 0x0,

        // Enumerate subdirectories
        Recurse = 0x1,

        // Skip files/directories when access is denied (e.g. AccessDeniedException/SecurityException)
        IgnoreInaccessible = 0x2,

        // Hint to use larger buffers for getting data (notably to help address remote enumeration perf)
        UseLargeBuffer = 0x4

        // Future: Add flags for tracking cycles, etc. 
    }
```

The new API surface area would be:

``` C#
    public static class Directory
    {
        public static IEnumerable<string> EnumerateDirectories(string path, string searchPattern, FindOptions findOptions);
        public static IEnumerable<string> EnumerateFiles(string path, string searchPattern, FindOptions findOptions);
        public static IEnumerable<string> EnumerateFileSystemEntries(string path, string searchPattern, FindOptions findOptions);
        public static string[] GetDirectories(string path, string searchPattern, FindOptions findOptions);
        public static string[] GetFiles(string path, string searchPattern, FindOptions findOptions);
        public static string[] GetFileSystemEntries(string path, string searchPattern, FindOptions findOptions);
    }

    public sealed class DirectoryInfo : FileSystemInfo
    {
        public IEnumerable<DirectoryInfo> EnumerateDirectories(string searchPattern, FindOptions findOptions);
        public IEnumerable<FileInfo> EnumerateFiles(string searchPattern, FindOptions findOptions);
        public IEnumerable<FileSystemInfo> EnumerateFileSystemInfos(string searchPattern, FindOptions findOptions);
        public DirectoryInfo[] GetDirectories(string searchPattern, FindOptions findOptions);
        public FileInfo[] GetFiles(string searchPattern, FindOptions findOptions);
        public FileSystemInfo[] GetFileSystemInfos(string searchPattern, FindOptions findOptions); 
    }
```

## Implementation Notes

- Won't check flags to allow light-up scenarios

**Comment by JeremyKuhne** [comment]
@terrajobst, @danmosemsft, @pjanotti

**Comment by danmosemsft** [comment]
typo `IgnoreInaccessable `

**Comment by JeremyKuhne** [comment]
> typo IgnoreInaccessable

Fixed, thanks

**Comment by jkotas** [comment]
>  Hint to use larger buffers for getting data (notably to help address remote enumeration perf)

How does one decide when to pass this flag? Should the name rather capture that, not what the implementation is going to do based on the flag?

> Allow .NET to skip locking if you know the enumerator won't be used on multiple threads

What is the performance benefit from this option? 

Note that there are examples where we have tried to add similar unsecure fast&crash options like these in the past (e.g. `bool isSynchronized` argument for `AppDomain.DefineDynamicAssembly`) that fired back and we have decide to just ignore the flag later.

**Comment by JeremyKuhne** [comment]
>>Hint to use larger buffers for getting data (notably to help address remote enumeration perf)

>How does one decide when to pass this flag? Should the name rather capture that, not what the implementation is going to do based on the flag?

By measurement. One known scenario is for remote shares, but there could be others. Not sure how better to name this, but I'm totally open to suggestions.

>> Allow .NET to skip locking if you know the enumerator won't be used on multiple threads

> What is the performance benefit from this option?

Looking again I didn't see a big enough benefit. Given your feedback I'll cull this option until I find harder evidence of substantive impact.

**Comment by jkotas** [comment]
> One known scenario is for remote shares, but there could be others. 

What I am going to lose if I set this flag all the time?

**Comment by JeremyKuhne** [comment]
> What I am going to lose if I set this flag all the time?

Just memory. The flag in Windows changes the buffer size from 4K to 64K. I'm inclined to not use 64K buffers by default (particularly from ArrayPool), but my gut feel may off. What do you think?

**Comment by jkotas** [comment]
Make sense. I cannot think about any better name.

**Comment by danmosemsft** [comment]
Can we identify at least some cases where we should probably default to `UseLargeBuffers`? Eg., if it is a UNC.

**Comment by terrajobst** [comment]
Superseded by #25873
